### PR TITLE
Support any number of strands from 1 to 8; make demo more like arduino demo

### LINKED ISSUE #5: Support num_strands=1
Over in #4, @gamblor21 identified that the 1 strand case is still not supported even with that PR. This is too bad, because the background write feature of NeoPxl8 increases framerate measurably.

To reduce memory usage and processing time I propose to add a third PIO program supporting a single strand in the original in-memory format. Still take a copy of the data for background writing without "tearing".

### PR DESCRIPTION
**Pull request by jepler**
Also closes #5

### CODE REVIEW

**Comment by jepler** [comment]
@gamblor21 do you want to test this PR with fewer than 8 pins?

**Comment by gamblor21** [comment]
> @gamblor21 do you want to test this PR with fewer than 8 pins?

Hope to test it tomorrow

**Review by gamblor21 — COMMENTED** [review]
Nice way of accounting for less then 8 strands. One issue though - `bitops.bit_tranpose` does not work for 1 strand (works 2-8).

I did test my setup (which is 1 strand) but setting num_strands to 2 and doubled the pixels and it did work. But either the library has to exclude 1 strand or bit_transpose needs a tweak or a 3rd PIO program that handles just 1 strand with a different data format.

Good news, filling 300 pixels (well 600 but half the fake ones) went from 73fps to about 182fps so I think getting it to work even for 1 strand may have worth.

**Comment by gamblor21** [comment]
I did manage to make a quick core change to `bitops.bit_transpose` to handle one strand. I can make a PR for it tomorrow (too late now) but basically in `shared-module/bitops/__init__.c` in `transpose_var` change the end of the switch to:
```C
        case 2:
            y |= *src << 8;
            src -= src_stride;
            MP_FALLTHROUGH;
        case 1:
            y |= *src;
```

and in the shared-bindings change the check to 1 instead of 2 as the minimum `num_strands`.

Also noticed there are some print statements still in the code.

**Comment by jepler** [comment]
You're not wrong to want support for 1 strip using background writes! Let's make that a separate issue, especially in case it requires a core change.

However, I think I'd rather have a separate program for 1 strip, because the extra memory usage and processing time for the bit_transpose operation can be eliminated entirely. A 2nd copy of the data is still needed in order to be free from tearing (writing a partially updated strip) but not a copy that's enlarged by 8x because each bit becomes a full byte! And it means no core change is required.

(a smaller savings could be had for 2 to 4 strands, with a core change, but I'm less convinced this is worth it)

**Comment by jepler** [comment]
I've filed an issue for the 1 strand case, so hopefully you can approve this PR as it stands.

File: adafruit_neopxl8.py [hunk 1]
```diff
@@ -135,11 +161,19 @@ def __init__(
 
         self._num_strands = num_strands
 
+        if num_strands == 8:
+            assembled = adafruit_pioasm.assemble(_PROGRAM8)
+        else:
+            program = _PROGRAMN.format(num_strands, 8 - num_strands)
+            print(program)
```
Comment (gamblor21) [inline adafruit_neopxl8.py:168]: Debugging print left behind. Also on line 171. Otherwise looks great.

**Comment by jepler** [comment]
I went ahead and added 1-strand support, but changed the structure a bit so it probably deserves a full re-test

I tested the 1-strand version on the neopixel bff.  The text scroller demo running unthrottled increases from 167 to 210fps! (and just commenting out `pixels.show` also gives 210fps :rofl: )

**Comment by jepler** [comment]
There was a mistake in my benchmarking of the neobff text scroller. The actual frame rates were 928fps (neopxl8 background) and 446fps (standard neopixel) :rocket:

**Review by gamblor21 — APPROVED** [review]
Tested with 1 and 2 strands, both worked great.

1 strand measure a speed-up from 73.2fps to 99.6fps over 300 neopixels. A 36% increase in speed.

**Pull request merged**